R/run-nmtran.R

```diff
@@ -0,0 +1,223 @@
+    ),
+    execute_nmtran(
+      nmtran_exe, mod_path = basename(mod_path), cmd_args = cmd_args,
+      dir = temp_folder
+    )
+  )
+
+  # assign class and return
+  class(nmtran_results) <- c(NMTRAN_PROCESS_CLASS, class(nmtran_results))
+  return(nmtran_results)
+}
+
+
+#' Search for and validate existence of an `NM-TRAN` executable
+#'
+#' If `nmtran_exe = NULL`, this will look for a `bbi.yaml` file in the same
+#' directory as the model.
+#'
+#' @inheritParams run_nmtran
+#'
+#' @keywords internal
+locate_nmtran <- function(.mod = NULL, .config_path = NULL, nmtran_exe = NULL){
+
+  if(is.null(nmtran_exe)){
+    if(!is.null(.mod)){
+      check_model_object(.mod, "bbi_nonmem_model")
+      model_dir <- get_model_working_directory(.mod)
+    }
+    config_path <- .config_path %||% file.path(model_dir, "bbi.yaml")
+
+    if(!file_exists(config_path)){
+      rlang::abort(
+        c(
+          "x" = "No bbi configuration was found in the execution directory.",
+          "i" = "Please run `bbi_init()` with the appropriate directory to continue."
+        )
+      )
+    }
+
+    if(!is.null(.config_path)){
+      config_path <- normalizePath(.config_path)
+    }
+
+    bbi_config <- yaml::read_yaml(config_path)
+    nm_config <- bbi_config$nonmem
+
+    # look for default nonmem installation
+    default_nm <- purrr::keep(nm_config, function(nm_ver){
+      !is.null(nm_ver$default)
+    })
+
+    # Set nonmem path
+    if(length(default_nm) > 0){
+      default_nm <- default_nm[[1]]
+    }else{
+      # If no default, use the last one (likely higher version)
+      default_nm <- nm_config[[length(nm_config)]]
+    }
+
+    # Set NMTRAN executable path
+    nm_path <- default_nm$home
+    # TODO: should we recursively look for this executable, or assume Metworx?
+    # i.e. can we assume NMTRAN is in a `tr` folder, and is called `NMTRAN.exe`?
+    nmtran_exe <- file.path(nm_path, "tr", "NMTRAN.exe")
+
+    # If executable found via bbi.yaml, append NONMEM version as attribute
+    attr(nmtran_exe, "nonmem_version") <- basename(default_nm$home)
+  }
+
+  if(!file_exists(nmtran_exe)){
+    stop(glue("Could not find an NMTRAN executable at `{nmtran_exe}`"))
+  }
+
+  return(nmtran_exe)
+}
+
+
+#' Execute `NM-TRAN` in a given directory
+#'
+#' @param nmtran_exe Path to `NM-TRAN` executable.
+#' @param mod_path Path of a model to evaluate. Should be relative to `dir`.
+#' @param cmd_args A character vector of command line arguments for the `NM-TRAN`
+#'  execution call
+#' @param dir Directory in which to execute the command.
+#'
+#' @keywords internal
+execute_nmtran <- function(nmtran_exe, mod_path, cmd_args = NULL, dir = NULL) {
+  if(is.null(dir)) dir <- "."
+  checkmate::assert_directory_exists(dir)
+
+  cmd_args <- if(is.null(cmd_args)) character() else cmd_args
+
+  nmtran.p <- processx::process$new(
+    command = nmtran_exe, wd = dir, args = cmd_args,
+    stdout = "|", stderr="|", stdin = file.path(dir, mod_path)
```

Comment:
As mentioned at <https://github.com/metrumresearchgroup/bbr/pull/705#discussion_r1677878012>, I think it'd be a good idea to preprocess this with nmtran_presort (when available) unless we're confident it doesn't matter.